common/src/tap/checks/value_check.rs

```diff
@@ -0,0 +1,535 @@
+// Copyright 2023-, GraphOps and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use ::cost_model::CostModel;
+use anyhow::anyhow;
+use bigdecimal::ToPrimitive;
+use sqlx::{
+    postgres::{PgListener, PgNotification},
+    PgPool,
+};
+use std::{
+    collections::HashMap,
+    str::FromStr,
+    sync::{Arc, RwLock},
+};
+use thegraph_core::DeploymentId;
+use tracing::error;
+
+use tap_core::receipt::{
+    checks::{Check, CheckError, CheckResult},
+    state::Checking,
+    Context, ReceiptWithState,
+};
+
+use crate::cost_model;
+
+// we only accept receipts with minimal 1 wei grt
+const MINIMAL_VALUE: u128 = 1;
+
+pub struct AgoraQuery {
+    pub deployment_id: DeploymentId,
+    pub query: String,
+    pub variables: String,
+}
+
+type CostModelMap = Arc<RwLock<HashMap<DeploymentId, CostModel>>>;
+type GlobalModel = Arc<RwLock<Option<CostModel>>>;
+
+pub struct MinimumValue {
+    cost_model_map: CostModelMap,
+    global_model: GlobalModel,
+    watcher_cancel_token: tokio_util::sync::CancellationToken,
+}
+
+struct CostModelWatcher {
+    pgpool: PgPool,
+
+    cost_models: CostModelMap,
+    global_model: GlobalModel,
+}
+
+impl CostModelWatcher {
+    async fn cost_models_watcher(
+        pgpool: PgPool,
+        mut pglistener: PgListener,
+        cost_models: CostModelMap,
+        global_model: GlobalModel,
+        cancel_token: tokio_util::sync::CancellationToken,
+    ) {
+        let cost_model_watcher = CostModelWatcher {
+            pgpool,
+            global_model,
+            cost_models,
+        };
+
+        loop {
+            tokio::select! {
+                _ = cancel_token.cancelled() => {
+                    break;
+                }
+                Ok(pg_notification) = pglistener.recv() => {
+                    cost_model_watcher.new_notification(
+                        pg_notification,
+                    ).await;
+                }
+            }
+        }
+    }
+
+    async fn new_notification(&self, pg_notification: PgNotification) {
```

Comment:
I think this function does too many things. It'd be better if it were broken up. For example,
* `parse_notification`
* `handle_insert`
* `handle_delete`
* `handle_unexpected_notification`

Not necessarily the best breakdown, or naming but just as an illustrative example.

Reply:
Much cleaner to me! Thanks!